control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+/*
```

Comment:
We should definitely come up with a better specification for how the table keys work. We can't ignore the types, though, because the keys are not specified using external names, but the header instance fields are. The fact that we have to translate between those two things is the reason that we have to walk the IR. You mention elsewhere that the inliner takes care of updating @name annotations, but I don't think that totally solves the problem because we still need to do with things like indexing header stacks. I would love to be wrong about that, though!

I do want to be clear that the *current* goal is to generate something that matches the existing BMV2 JSON -> PI pathway pretty closely. This makes it much easier to compare the two pathways, which is very handy during development. More importantly, it also makes it feasible for users to migrate from one to the other. Keep in mind that the ids we use at the control plane API level are calculated from the names, so we can't change the names without changing the ids, which will break things for users. Until the BMV2 JSON -> PI pathway is deprecated (which is a place we want to get to ASAP), we've really got to be careful to keep them compatible.

I'd love to talk about this more with you Mihai. We're very much on the same page about making this generic. I'm just trying to get us there incrementally so we can maintain compatibility.